gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/AbstractRexster.java

```diff
@@ -0,0 +1,758 @@
+            final String errorMessage = String.format("The Gremlin statement that was submitted exceeds the maximum compilation size allowed by the JVM, please split it into multiple smaller statements - %s", trimMessage(gremlinContext.getRequestMessage()));
+            logger.warn(errorMessage);
+            throw new RexsterException(errorMessage, root, ResponseMessage.build(gremlinContext.getRequestMessage())
+                    .code(ResponseStatusCode.SERVER_ERROR_EVALUATION)
+                    .statusMessage(errorMessage)
+                    .statusAttributeException(root).create());
+        }
+
+        // GroovyRuntimeException will hit a pretty wide range of eval type errors, like MissingPropertyException,
+        // CompilationFailedException, MissingMethodException, etc. If more specific handling is required then
+        // try to catch it earlier above.
+        if (root instanceof GroovyRuntimeException ||
+                root instanceof VerificationException ||
+                root instanceof ScriptException) {
+            throw new RexsterException(root.getMessage(), root, ResponseMessage.build(gremlinContext.getRequestMessage())
+                    .code(ResponseStatusCode.SERVER_ERROR_EVALUATION)
+                    .statusMessage(root.getMessage())
+                    .statusAttributeException(root).create());
+        }
+
+        throw new RexsterException(root.getClass().getSimpleName() + ": " + root.getMessage(), root,
+                ResponseMessage.build(gremlinContext.getRequestMessage())
+                        .code(ResponseStatusCode.SERVER_ERROR)
+                        .statusAttributeException(root)
+                        .statusMessage(root.getMessage()).create());
+    }
+
+    /**
+     * Used to decrease the size of a Gremlin script that triggered a "method too large" exception so that it
+     * doesn't log a massive text string nor return a large error message.
+     */
+    private RequestMessage trimMessage(final RequestMessage msg) {
+        final RequestMessage trimmedMsg = RequestMessage.from(msg).create();
+        if (trimmedMsg.getArgs().containsKey(Tokens.ARGS_GREMLIN))
+            trimmedMsg.getArgs().put(Tokens.ARGS_GREMLIN, trimmedMsg.getArgs().get(Tokens.ARGS_GREMLIN).toString().substring(0, 1021) + "...");
+
+        return trimmedMsg;
+    }
+
+    /**
+     * Check if any exception in the chain is TemporaryException then we should respond with the right error code so
+     * that the client knows to retry.
+     */
+    protected Optional<Throwable> determineIfTemporaryException(final Throwable ex) {
+        return Stream.of(ExceptionUtils.getThrowables(ex)).
+                filter(i -> i instanceof TemporaryException).findFirst();
+    }
+
+    @Override
+    public synchronized void close() {
+        // already closing/closed
+        if (!sessions.containsKey(sessionId)) return;
```

Comment:
how do we handle this case:
1. request1 for sessionA gets enqueued and starts processing
2. request2 for sessionA gets enqueued and hasn't started processing yet
3. client closes session
4. request1 gets interrupted and rolledback. session ID is removed from the map
5. request2 is dequeued, it tried to find if a sessionA exists and it finds that it doesn't so it creates another one and executes successfully.

From a client perspective, they sent 2 requests for a session and then closed the session which should have rolledback both of them but now client observes that request2 is successful. We can say that request2 will not commit ever unless we get a commit from the client for sessionA but it still ends up using resources to execute.

Instead of a Map, I think we should be using  DS where we can evict the closed sessions after some amount of time i.e. we still keep their IDs in the DS even when closed. If a request arrives for a closed session we throw exception. Since we don't want this DS to increase indefinitely, we should evict closed entries after some time. Ideally this DS should evict closed session after X number of requests have been dequeued where X is the number of requests in queue when the session was closed.

Reply:
If request2 has made it to the queue when request1 was interrupted, then it will be removed from the queue with error returned to the client.  It will not use resources to execute. I think this test is showing the scenario you're talking about.

https://github.com/apache/tinkerpop/commit/04a3b9a9db2008be82e4593f5b9ae4a16a4d4b81

Reply:
> Instead of a Map, I think we should be using DS where we can evict the closed sessions after some amount of time i.e. we still keep their IDs in the DS even when closed. If a request arrives for a closed session we throw exception. Since we don't want this DS to increase indefinitely, we should evict closed entries after some time. Ideally this DS should evict closed session after X number of requests have been dequeued where X is the number of requests in queue when the session was closed.

maybe my mind is blanking but what is a "DS"? i'm going to feel dumb when you say what that is but i considered some kind of system like the one you describe to keep the ids but then it wouldn't let us replicate the original functionality of sessions where the id could be re-used, which would prevent applications from working the way they always have. i wasnt sure if that was a step we should be taking.